### DISCUSSION TITLE
Request: Support for Proxy Protocol

### DISCUSSION THREAD

**Discussion by mikker** [basecamp/kamal-proxy#41] [Ideas and Issue Triage]
I run my Kamal hosted app behind a load balancer. Previously, to preserve the user's original IP address, I set up Traefik to support [Proxy Protocol](https://doc.traefik.io/traefik/routing/entrypoints/#proxyprotocol) which Digital Ocean also supports.

Is this something that's on the planning board for kamal-proxy? Or is the suggested remedy to put traefik in between again?

**Comment by eminkel** [comment]
Looking for a similar feature.

Getting the improper IP in the application logs (outer most proxy). 

I was able to configure in the past as found here: https://github.com/basecamp/kamal/issues/576#issuecomment-1926269676

**Reply by eminkel** [reply]
I was able to configure this issue away at the application level. I'd much rather manage it at the proxy, but this will move me forward.

https://unfit-for.work/posts/2022/gcp-rails-remote-ip/

**Comment by kevinmcconnell** [comment]
We don't have Proxy Protocol support at the moment, but I'll look into it.

Kamal Proxy does allow populating/forwarding `X-Forwarded-For`, which is another common way to pass through the original IP. Slightly different approach, but can often be a good alternative.

**Comment by chrispage1** [comment]
I agree send-proxy would be really helpful. We run haproxy as our load balancers so routing via TCP through to the Kamal proxy hosts without losing IP information would be great.

**Comment by jlecour** [comment]
Hi,

Using `X-Forwarded-For` is indeed the simplest way to deal with the "original IP" situation, but it does it at the OSI level 7 (application layer) and has a lot disadvantages.
Using the PROXY protocol achives the same result at the OSI level 4 (transport layer) and is invisible to the layer 7, which is great.

I think it would be great if Kamal supported both options.

**Comment by basiszwo** [comment]
my setup is, that I need to have a load balancer for multiple applications running behind this load balancer. the LB does not terminate SSL and passes it onto the backends which run a standard kamal setup. 

The problem is, that the `X-Forwarded-For` header cannot be injected as we're dealing with layer 4. that's where the proxy protocol comes in. 

If kamal could support this, that'll be super nice as other solutions involve running traefik instead of / in addition to kamal-proxy or having an nginx running as an accessory.

**Reply by basiszwo** [reply]
I went ahead and implemented PROXY protocol support for kamal-proxy. Happy to open a PR if there's interest.

Implementation summary:
  - Uses https://github.com/pires/go-proxyproto which supports both v1 and v2
  - Adds --proxy-protocol flag (and PROXY_PROTOCOL env var)
  - Wraps the HTTP/HTTPS listeners with proxyproto.Listener
  - The wrapped listener parses the PROXY header on Accept() and makes the real client IP available via RemoteAddr()
  - No changes needed to the logging middleware - r.RemoteAddr automatically contains the real client IP

Files changed:
  - go.mod - add dependency
  - internal/server/config.go - add config field
  - internal/cmd/run.go - add CLI flag
  - internal/server/server.go - wrap listeners

Security note: PROXY protocol should only be enabled when running behind a trusted load balancer. If exposed directly to the internet, clients could spoof their source IP.

Let me know if you'd like me to open a PR, or if there are any changes you'd prefer to the approach.
